indexer/src/main/kotlin/com/openlattice/BackgroundExternalDatabaseSyncingService.kt

```diff
@@ -238,48 +177,137 @@ class BackgroundExternalDatabaseSyncingService(
+        val tableCols = edms.getColumnMetadata(table)
+        val tableColNames = tableCols.map { it.getUniqueName() }.toSet()
+
+        val existingColumnIdsByName = reservationService.getIdsByFqn(tableColNames)
+
+        return tableCols.groupBy { existingColumnIdsByName.contains(it.getUniqueName()) }.flatMap { (shouldUpdate, cols) ->
+            if (cols.isEmpty()) {
+                return@flatMap listOf<OrganizationExternalDatabaseColumn>()
+            }
+
+            if (shouldUpdate) {
+                updateColumns(cols, existingColumnIdsByName)
+            } else {
+                createColumns(table, cols, ownerAclKeys)
+            }
+        }.map { it.id }.toSet()
+    }
+
+    private fun removeNonexistentTablesAndColumnsForOrg(orgId: UUID, existingTableIds: Set<UUID>, existingColumnIds: Set<UUID>) {
+
+        // delete missing tables
+
+        val tableIdsToDelete = organizationExternalDatabaseTables.keySet(Predicates.equal(ORGANIZATION_ID_INDEX, orgId)).filter {
+            !existingTableIds.contains(it)
+        }.toSet()
+
+        if (tableIdsToDelete.isNotEmpty()) {
+            edms.deleteOrganizationExternalDatabaseTableObjects(tableIdsToDelete)
+        }
+
+
+        // delete missing columns
+
+        val columnIdsToDelete = organizationExternalDatabaseColumns.values(Predicates.equal(ORGANIZATION_ID_INDEX, orgId)).filter {
+            !existingColumnIds.contains(it.id)
+        }.groupBy { it.tableId }.mapValues { it.value.map { c -> c.id }.toSet() }
+
+        if (columnIdsToDelete.isNotEmpty()) {
+            edms.deleteOrganizationExternalDatabaseColumnObjects(columnIdsToDelete)
+        }
+    }
+
+    private fun createColumns(
+            table: OrganizationExternalDatabaseTable,
+            columns: List<OrganizationExternalDatabaseColumn>,
+            ownerAclKeys: List<AclKey>
+    ): List<OrganizationExternalDatabaseColumn> {
+        createSecurableColumnObjects(columns, ownerAclKeys, table)
+
+        return columns
+    }
+
+    private fun updateColumns(
+            columns: List<OrganizationExternalDatabaseColumn>,
+            existingColumnIdsByName: Map<String, UUID>
+    ): List<OrganizationExternalDatabaseColumn> {
+
+        val currentColumnsById = columns.associateBy { existingColumnIdsByName.getValue(it.getUniqueName()) }
+        val storedColumns = organizationExternalDatabaseColumns.getAll(existingColumnIdsByName.values.toSet())
+
+        val fullExistingColumns = mutableListOf<OrganizationExternalDatabaseColumn>()
+
+        val updatedColumns = storedColumns.mapNotNull { (id, storedColumn) ->
+            val currentColumn = currentColumnsById.getValue(id)
+            val updatedColumn = OrganizationExternalDatabaseColumn(
+                    id = storedColumn.id,
+                    name = currentColumn.name,
+                    title = storedColumn.title,
```

Comment:
What's the logic for using `currentColumn` vs `storedColumn`? `name` should be immutable and `title` is mutable, right? In which case, shouldn't it be `name = storedColumn.name` and `title = currentColumn.title`?

Reply:
no, the goal here is to not overwrite any metadata users have added -- if they add a description for a column through the UI, and that column is synced, we don't want that description to disappear.